saleor/attribute/migrations/0020_auto_20220214_1027.py

```diff
@@ -0,0 +1,122 @@
+    attr_values_to_delete = set()
+
+    attr_value_map = defaultdict(list)
+    for attr_value in attr_values_qs:
+
+        ref_id = fetch_reference_id_from_slug(attr_value.slug)
+        attr_value_map[ref_id].append(attr_value)
+
+    model_queryset = Model.objects.filter(pk__in=attr_value_map.keys())
+    model_in_bulk = model_queryset.in_bulk()
+
+    model_ids = set(model_queryset.values_list("id", flat=True))
+    ids_to_delete = set(attr_value_map.keys()) - model_ids
+    ids_to_update = set(attr_value_map.keys()) - set(ids_to_delete)
+
+    for id_to_delete in ids_to_delete:
+        attr_values = attr_value_map[id_to_delete]
+        for attr_value in attr_values:
+            attr_values_to_delete.add(attr_value.id)
+
+    if Model.__name__ == "Page":
+        for id_to_update in ids_to_update:
+            attr_values = attr_value_map[id_to_update]
+            for attr_value in attr_values:
+                attr_value.reference_page = model_in_bulk[id_to_update]
+            attr_values_to_update.extend(attr_values)
+
+    if Model.__name__ == "Product":
+        for id_to_update in ids_to_update:
+            attr_values = attr_value_map[id_to_update]
+            for attr_value in attr_values:
+                attr_value.reference_product = model_in_bulk[id_to_update]
+            attr_values_to_update.extend(attr_values)
+
+    return attr_values_to_update, attr_values_to_delete
+
+
+def migrate_model_field_data(apps, schema):
+    AttributeValue = apps.get_model("attribute", "AttributeValue")
+    Attribute = apps.get_model("attribute", "Attribute")
+    Page = apps.get_model("page", "Page")
+    Product = apps.get_model("product", "Product")
+
+    queryset = AttributeValue.objects.filter(
+        Exists(
+            Attribute.objects.filter(
+                id=OuterRef("attribute_id"), input_type="reference", entity_type="Page"
+            )
+        )
+    )
+    for batch_pks in queryset_in_batches(queryset):
+        batch = AttributeValue.objects.filter(pk__in=batch_pks)
+        attr_values_to_update, attr_values_to_delete = import_attribute_values(
+            batch, Page
+        )
+        AttributeValue.objects.bulk_update(attr_values_to_update, ["reference_page"])
+        AttributeValue.objects.filter(pk__in=attr_values_to_delete).delete()
+
+    queryset = AttributeValue.objects.filter(
+        Exists(
+            Attribute.objects.filter(
+                id=OuterRef("attribute_id"),
+                input_type="reference",
+                entity_type="Product",
+            )
+        )
+    )
+    for batch_pks in queryset_in_batches(queryset):
+        batch = AttributeValue.objects.filter(pk__in=batch_pks)
+        attr_values_to_update, attr_values_to_delete = import_attribute_values(
+            batch, Product
+        )
+        AttributeValue.objects.bulk_update(attr_values_to_update, ["reference_product"])
+        AttributeValue.objects.filter(pk__in=attr_values_to_delete).delete()
```

Comment:
These two parts look similar. IMO we can create one generic function for that, or at least separate it into two different methods.